File: android2_lesson01/src/java/java/com/example/ama/android2_lesson01/model/Note.java
package com.example.ama.android2_lesson01.model;

import android.os.Parcel;
import android.os.Parcelable;

import com.example.ama.android2_lesson01.NotesApp;
import com.example.ama.android2_lesson01.R;

/**
 * Class describing notes
 */

public class Note implements Parcelable {
    private long mId;
    private String mTitle;
    private String mText;

    /**
     * Constructor
     *
     * @param mId    note id from the database
     * @param mTitle note title
     * @param mText  note text
     */

    public Note(long mId, String mTitle, String mText) {
        this.mId = mId;
        this.mTitle = mTitle;
        this.mText = mText;
    }

    protected Note(Parcel in) {
        mId = in.readLong();
        mTitle = in.readString();
        mText = in.readString();
    }

    /**
     * Get new instance on NoteBuilder for Note creation
     *
     * @return new NoteBuilder instance
     */

    public static NoteBuilder builder() {
        return new NoteBuilder();
    }

    public static final Creator<Note> CREATOR = new Creator<Note>() {
        @Override
        public Note createFromParcel(Parcel in) {
            return new Note(in);
        }

        @Override
        public Note[] newArray(int size) {
            return new Note[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(mId);
        dest.writeString(mTitle);
        dest.writeString(mText);
    }

    public String getTitle() {
        return mTitle;
    }

    public String getText() {
        return mText;
    }

    public long getmId() {
        return mId;
    }

    /**
     * Class for Note building
     */

    public static class NoteBuilder {
        private long mId = 0;
        private String mTitle = NotesApp.getInstance().getString(R.string.default_note_title);
        private String mText = NotesApp.getInstance().getString(R.string.default_note_text);

        public NoteBuilder id(long mId) {
            this.mId = mId;
            return this;
        }

        public NoteBuilder title(String mTitle) {
            if (mTitle != null && !mTitle.equals("")) {
                this.mTitle = mTitle;
            }
            return this;
        }

        public NoteBuilder text(String mText) {
            if (mText != null && !mText.equals("")) {
                this.mText = mText;
            }
            return this;
        }

        public Note build() {
            return new Note(mId, mTitle, mText);
        }
    }
}

File: android2_lesson01/src/kotlin/java/com/example/ama/android2_lesson01/model/Note.kt
package com.example.ama.android2_lesson01.model

import android.os.Parcel
import android.os.Parcelable
import com.example.ama.android2_lesson01.NotesApp
import com.example.ama.android2_lesson01.R

/**
 * Class describing notes
 */

class Note(val id: Long,
           val title: String,
           val text: String)
    : Parcelable {

    constructor(parcel: Parcel) : this(
            parcel.readLong(),
            parcel.readString(),
            parcel.readString()) {
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeLong(id)
        parcel.writeString(title)
        parcel.writeString(text)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Note> {

        /**
         * Get new instance on NoteBuilder for Note creation
         *
         * @return new NoteBuilder instance
         */

        fun builder(): NoteBuilder = NoteBuilder()

        override fun createFromParcel(parcel: Parcel): Note {
            return Note(parcel)
        }

        override fun newArray(size: Int): Array<Note?> {
            return arrayOfNulls(size)
        }
    }

    /**
     * Class for Note building
     */

    class NoteBuilder {
        var id: Long = 0
            private set
        var title = NotesApp.instance.resources.getString(R.string.default_note_title)!!
            private set
        var text = NotesApp.instance.resources.getString(R.string.default_note_text)!!
            private set

        fun id(id: Long): NoteBuilder {
            this.id = id
            return this
        }

        fun title(title: String): NoteBuilder {
            if (title != "") {
                this.title = title
            }
            return this
        }

        fun text(text: String): NoteBuilder {
            if (text != "") {
                this.text = text
            }
            return this
        }

        fun build() = Note(id, title, text)
    }
}